{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "5e9951cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os.path import isfile\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.cm import ScalarMappable\n",
    "import textwrap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6eeb8f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path = \"../../results/gene_enrichment_analysis/goatools/\"\n",
    "data_path = \"../../results/rsem/\"\n",
    "samples_path = \"../../data/rsem/samples.csv\"\n",
    "samples_dataframe = pd.read_csv(samples_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc1b400",
   "metadata": {},
   "source": [
    "# Preparation of input files for goatools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "957dfd3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../../results/rsem/liquid_mono_culture_orgint_vs_metatranscriptome.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_orgint_vs_plate_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_orgint_vs_liquid_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_orgint_vs_hydra_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/metatranscriptome_vs_liquid_mono_culture_orgint.CSV',\n",
       " '../../results/rsem/metatranscriptome_vs_plate_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/metatranscriptome_vs_liquid_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/metatranscriptome_vs_hydra_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/plate_mono_culture_kiel_vs_liquid_mono_culture_orgint.CSV',\n",
       " '../../results/rsem/plate_mono_culture_kiel_vs_metatranscriptome.CSV',\n",
       " '../../results/rsem/plate_mono_culture_kiel_vs_liquid_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/plate_mono_culture_kiel_vs_hydra_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_kiel_vs_liquid_mono_culture_orgint.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_kiel_vs_metatranscriptome.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_kiel_vs_plate_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/liquid_mono_culture_kiel_vs_hydra_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/hydra_mono_culture_kiel_vs_liquid_mono_culture_orgint.CSV',\n",
       " '../../results/rsem/hydra_mono_culture_kiel_vs_metatranscriptome.CSV',\n",
       " '../../results/rsem/hydra_mono_culture_kiel_vs_plate_mono_culture_kiel.CSV',\n",
       " '../../results/rsem/hydra_mono_culture_kiel_vs_liquid_mono_culture_kiel.CSV']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes = []\n",
    "for unique_sample in samples_dataframe.treatment.unique():\n",
    "    for unique_sample_2 in samples_dataframe.treatment.unique():\n",
    "        if unique_sample != unique_sample_2:\n",
    "            file = data_path + unique_sample + \"_vs_\" + unique_sample_2 + \".CSV\"\n",
    "            if isfile(file):\n",
    "                dataframes.append(file)\n",
    "dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7ec4f6d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>wp_number</th>\n",
       "      <th>description</th>\n",
       "      <th>GO</th>\n",
       "      <th>GO_process</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AEP_RS00005</td>\n",
       "      <td>gene:AEP_00001</td>\n",
       "      <td>WP_087493495.1</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>GO:0000160,GO:0006355,GO:0003677</td>\n",
       "      <td>phosphorelay signal transduction system|000016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>AEP_RS00010</td>\n",
       "      <td>gene:AEP_00002</td>\n",
       "      <td>WP_087493496.1</td>\n",
       "      <td>sodium-translocating pyrophosphatase</td>\n",
       "      <td>GO:1902600,GO:0009678</td>\n",
       "      <td>proton transmembrane transport|1902600||IEA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AEP_RS00015</td>\n",
       "      <td>gene:AEP_00003</td>\n",
       "      <td>WP_087493497.1</td>\n",
       "      <td>inorganic diphosphatase</td>\n",
       "      <td>GO:0006796,GO:0004427</td>\n",
       "      <td>phosphate-containing compound metabolic proces...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AEP_RS00020</td>\n",
       "      <td>gene:AEP_00004</td>\n",
       "      <td>WP_087493498.1</td>\n",
       "      <td>alpha/beta fold hydrolase</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AEP_RS00025</td>\n",
       "      <td>gene:AEP_00005</td>\n",
       "      <td>WP_087493499.1</td>\n",
       "      <td>chemotaxis protein CheW</td>\n",
       "      <td>GO:0006935,GO:0007165</td>\n",
       "      <td>chemotaxis|0006935||IEA,signal transduction|00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0    locus_tag         gene_id       wp_number  \\\n",
       "0           0  AEP_RS00005  gene:AEP_00001  WP_087493495.1   \n",
       "1           1  AEP_RS00010  gene:AEP_00002  WP_087493496.1   \n",
       "2           2  AEP_RS00015  gene:AEP_00003  WP_087493497.1   \n",
       "3           3  AEP_RS00020  gene:AEP_00004  WP_087493498.1   \n",
       "4           4  AEP_RS00025  gene:AEP_00005  WP_087493499.1   \n",
       "\n",
       "                               description                                GO  \\\n",
       "0  response regulator transcription factor  GO:0000160,GO:0006355,GO:0003677   \n",
       "1     sodium-translocating pyrophosphatase             GO:1902600,GO:0009678   \n",
       "2                  inorganic diphosphatase             GO:0006796,GO:0004427   \n",
       "3                alpha/beta fold hydrolase                           unknown   \n",
       "4                  chemotaxis protein CheW             GO:0006935,GO:0007165   \n",
       "\n",
       "                                          GO_process  \n",
       "0  phosphorelay signal transduction system|000016...  \n",
       "1        proton transmembrane transport|1902600||IEA  \n",
       "2  phosphate-containing compound metabolic proces...  \n",
       "3                                            unknown  \n",
       "4  chemotaxis|0006935||IEA,signal transduction|00...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curvibacter_genes_df = pd.read_csv(\"../../results/curvibacter_genome_annotation.csv\")\n",
    "curvibacter_genes_df[\"gene_id\"] = curvibacter_genes_df[\"gene_id\"].apply(lambda x: \"gene:\"+x)\n",
    "curvibacter_genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "18e13f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing population and associations file \n",
    "go_gene_ids = []\n",
    "with open(result_path + \"associations.txt\", \"w\") as associations_file:\n",
    "    with open(result_path + \"populations.txt\", \"w\") as populations_file:\n",
    "        for gene_id in curvibacter_genes_df[\"gene_id\"]:\n",
    "            if curvibacter_genes_df[curvibacter_genes_df[\"gene_id\"] == gene_id].GO.values[0] != \"unknown\":\n",
    "                \n",
    "                go_gene_ids.append(gene_id)\n",
    "                populations_file.write(gene_id+\"\\n\")\n",
    "                associations_file.write(gene_id + \"\\t\")\n",
    "                counter = 0\n",
    "                gos = curvibacter_genes_df[curvibacter_genes_df[\"gene_id\"] == gene_id].GO.values[0].split(\",\")\n",
    "                for go_id in gos:\n",
    "                    \n",
    "                    if counter != len(gos)-1:\n",
    "                        associations_file.write(go_id+\";\")\n",
    "                    else:\n",
    "                        associations_file.write(go_id+\"\\n\")\n",
    "                    counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "eca183f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[+] Working with: ../../results/rsem/liquid_mono_culture_orgint_vs_metatranscriptome.CSV\n",
      "\t[*] Length of downregulated genes: 325\n",
      "\t[*] Length of upregulated genes: 344\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_orgint_vs_plate_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 344\n",
      "\t[*] Length of upregulated genes: 331\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_orgint_vs_liquid_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 221\n",
      "\t[*] Length of upregulated genes: 270\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_orgint_vs_hydra_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 340\n",
      "\t[*] Length of upregulated genes: 531\n",
      "[+] Working with: ../../results/rsem/metatranscriptome_vs_liquid_mono_culture_orgint.CSV\n",
      "\t[*] Length of downregulated genes: 344\n",
      "\t[*] Length of upregulated genes: 325\n",
      "[+] Working with: ../../results/rsem/metatranscriptome_vs_plate_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 501\n",
      "\t[*] Length of upregulated genes: 489\n",
      "[+] Working with: ../../results/rsem/metatranscriptome_vs_liquid_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 387\n",
      "\t[*] Length of upregulated genes: 419\n",
      "[+] Working with: ../../results/rsem/metatranscriptome_vs_hydra_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 53\n",
      "\t[*] Length of upregulated genes: 267\n",
      "[+] Working with: ../../results/rsem/plate_mono_culture_kiel_vs_liquid_mono_culture_orgint.CSV\n",
      "\t[*] Length of downregulated genes: 331\n",
      "\t[*] Length of upregulated genes: 344\n",
      "[+] Working with: ../../results/rsem/plate_mono_culture_kiel_vs_metatranscriptome.CSV\n",
      "\t[*] Length of downregulated genes: 489\n",
      "\t[*] Length of upregulated genes: 501\n",
      "[+] Working with: ../../results/rsem/plate_mono_culture_kiel_vs_liquid_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 255\n",
      "\t[*] Length of upregulated genes: 322\n",
      "[+] Working with: ../../results/rsem/plate_mono_culture_kiel_vs_hydra_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 346\n",
      "\t[*] Length of upregulated genes: 531\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_kiel_vs_liquid_mono_culture_orgint.CSV\n",
      "\t[*] Length of downregulated genes: 270\n",
      "\t[*] Length of upregulated genes: 221\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_kiel_vs_metatranscriptome.CSV\n",
      "\t[*] Length of downregulated genes: 419\n",
      "\t[*] Length of upregulated genes: 387\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_kiel_vs_plate_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 322\n",
      "\t[*] Length of upregulated genes: 255\n",
      "[+] Working with: ../../results/rsem/liquid_mono_culture_kiel_vs_hydra_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 342\n",
      "\t[*] Length of upregulated genes: 475\n",
      "[+] Working with: ../../results/rsem/hydra_mono_culture_kiel_vs_liquid_mono_culture_orgint.CSV\n",
      "\t[*] Length of downregulated genes: 531\n",
      "\t[*] Length of upregulated genes: 340\n",
      "[+] Working with: ../../results/rsem/hydra_mono_culture_kiel_vs_metatranscriptome.CSV\n",
      "\t[*] Length of downregulated genes: 267\n",
      "\t[*] Length of upregulated genes: 53\n",
      "[+] Working with: ../../results/rsem/hydra_mono_culture_kiel_vs_plate_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 531\n",
      "\t[*] Length of upregulated genes: 346\n",
      "[+] Working with: ../../results/rsem/hydra_mono_culture_kiel_vs_liquid_mono_culture_kiel.CSV\n",
      "\t[*] Length of downregulated genes: 475\n",
      "\t[*] Length of upregulated genes: 342\n"
     ]
    }
   ],
   "source": [
    "# writing sample files for up/down regulated genes\n",
    "go_files = []\n",
    "for df in dataframes:\n",
    "    print(\"[+] Working with: {}\".format(df))\n",
    "    log2folddf = pd.read_csv(df)\n",
    "    log2folddf.columns = [\"gene_id\",\"baseMean\",\"log2FoldChange\",\"lfcSE\",\"stat\",\"pvalue\",\"padj\"]\n",
    "    log2folddf = log2folddf[log2folddf[\"padj\"] <= 0.05]\n",
    "    downregulated_genes = log2folddf[log2folddf[\"log2FoldChange\"] <= -1.0]\n",
    "    upregulated_genes = log2folddf[log2folddf[\"log2FoldChange\"] >= 1.0]\n",
    "\n",
    "    downregulated_genes = downregulated_genes[downregulated_genes[\"gene_id\"].isin(go_gene_ids)]\n",
    "    upregulated_genes = upregulated_genes[upregulated_genes[\"gene_id\"].isin(go_gene_ids)]\n",
    "    \n",
    "    \n",
    "    print(\"\\t[*] Length of downregulated genes: {}\".format(len(downregulated_genes)))\n",
    "    print(\"\\t[*] Length of upregulated genes: {}\".format(len(upregulated_genes)))\n",
    "    \n",
    "    sample = df.split(\"/\")[-1].split(\".CSV\")[0] \n",
    "    sample_up = result_path + sample + \"_upregulated_genes_goatools.txt\"\n",
    "    sample_down = result_path + sample + \"_downregulated_genes_goatools.txt\"\n",
    "    \n",
    "    if len(downregulated_genes) > 5:\n",
    "        go_files.append(sample_down)\n",
    "        with open(sample_down, \"w\") as goadown:\n",
    "            for gene_id in downregulated_genes[\"gene_id\"]:\n",
    "                goadown.write(gene_id+\"\\n\")\n",
    "    if len(upregulated_genes) > 5:\n",
    "        go_files.append(sample_up)\n",
    "        with open(sample_up, \"w\") as goaup:\n",
    "            for gene_id in upregulated_genes[\"gene_id\"]:\n",
    "                goaup.write(gene_id+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "ec1584cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_metatranscriptome_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_metatranscriptome_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_plate_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_plate_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_liquid_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_liquid_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_hydra_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_orgint_vs_hydra_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_liquid_mono_culture_orgint_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_liquid_mono_culture_orgint_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_plate_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_plate_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_liquid_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_liquid_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_hydra_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/metatranscriptome_vs_hydra_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_liquid_mono_culture_orgint_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_liquid_mono_culture_orgint_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_metatranscriptome_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_metatranscriptome_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_liquid_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_liquid_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_hydra_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/plate_mono_culture_kiel_vs_hydra_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_liquid_mono_culture_orgint_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_liquid_mono_culture_orgint_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_metatranscriptome_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_metatranscriptome_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_plate_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_plate_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_hydra_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/liquid_mono_culture_kiel_vs_hydra_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_liquid_mono_culture_orgint_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_liquid_mono_culture_orgint_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_metatranscriptome_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_metatranscriptome_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_plate_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_plate_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_liquid_mono_culture_kiel_downregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n",
      "[+] Working with ../../results/gene_enrichment_analysis/goatools/hydra_mono_culture_kiel_vs_liquid_mono_culture_kiel_upregulated_genes_goatools.txt\n",
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "[*] DONE\n"
     ]
    }
   ],
   "source": [
    "goafiles = []\n",
    "for samplefile in go_files:\n",
    "    print(\"[+] Working with {}\".format(samplefile))\n",
    "    outfile = result_path + samplefile.split(\"/\")[-1].split(\".txt\")[0] + \"_output.table\"\n",
    "    \n",
    "    !find_enrichment.py $samplefile ../../results/gene_enrichment_analysis/goatools/populations.txt ../../results/gene_enrichment_analysis/goatools/associations.txt --annofmt id2gos --alpha 0.05 --pval 0.05 --obo ../../results/gene_enrichment_analysis/goatools/go-basic.obo --method fdr_bh --outfile $outfile --obsolete replace > /dev/null\n",
    "    \n",
    "    goafiles.append(outfile)\n",
    "    print(\"[*] DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d519b40d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_goa(goafile_enriched:pd.DataFrame,savep:str, filename:str):\n",
    "    print(\"[*] Producing plot for {}\".format(filename))\n",
    "    goafile_enriched[\"ratio_stud\"] = goafile_enriched.ratio_in_study.apply(lambda x: int(x.split(\"/\")[0])/int(x.split(\"/\")[1]))\n",
    "    goafile_enriched[\"ratio_pop\"] = goafile_enriched.ratio_in_pop.apply(lambda x: int(x.split(\"/\")[0])/int(x.split(\"/\")[1]))\n",
    "    goafile_enriched[\"amount_in_pop\"] = goafile_enriched.ratio_in_pop.apply(lambda x: int(x.split(\"/\")[0]))\n",
    "\n",
    "    categories = []\n",
    "    for cat in list(goafile_enriched.name):\n",
    "        if len(cat) >= 30:\n",
    "            cat = textwrap.fill(cat, width=30)\n",
    "        categories.append(cat)\n",
    "        \n",
    "    values = list(goafile_enriched.study_count)\n",
    "    scatter_values = np.array(goafile_enriched.study_count) / np.array(goafile_enriched.amount_in_pop)\n",
    "    \n",
    "    pcolors = goafile_enriched.p_fdr_bh\n",
    "    norm_p_values = np.array(pcolors) / max(pcolors)\n",
    "    colors=plt.cm.RdBu_r(norm_p_values)\n",
    "\n",
    "    # Create figure and axes\n",
    "    if len(goafile_enriched) == 30:\n",
    "        fsize = (20,18)\n",
    "    elif len(goafile_enriched) >= 15:\n",
    "        fsize = (16,12)\n",
    "    else:\n",
    "        fsize = (12,8)\n",
    "        \n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 16), sharey=True)\n",
    "\n",
    "    # Plot horizontal bar plot on ax1\n",
    "    ax1.barh(categories, values, color=colors, edgecolor=\"black\")\n",
    "\n",
    "    ax1.set_xlabel('Count')\n",
    "\n",
    "    ax2.scatter(scatter_values, categories, c=colors, cmap='RdBu_r', \n",
    "                label='Gene Ratio (compared to Study)', s=list(goafile_enriched.ratio_stud*1000),edgecolor=\"black\")\n",
    "\n",
    "    ax2.set_xlabel('Count in Study / Count in Pop')\n",
    "    ax1.set_ylabel('GO Categories')\n",
    "    ax1.invert_yaxis()\n",
    "    plt.subplots_adjust(left=0.2, wspace=0.1)\n",
    "    cbar = fig.colorbar(ScalarMappable(cmap='RdBu_r'), ax=[ax1, ax2], pad = 0.005)\n",
    "    cbar.set_label('P-values')\n",
    "    cbar.set_ticks([min(norm_p_values), max(norm_p_values)])\n",
    "    cbar.set_ticklabels([f'{min(goafile_enriched.p_fdr_bh):.4f}', f'{max(goafile_enriched.p_fdr_bh):.4f}'])\n",
    "\n",
    "    cbar.ax.set_position([0.85, 0.15, 0.03, 0.7])\n",
    "\n",
    "    plt.savefig(savep + filename + \".jpg\", dpi=400)\n",
    "    plt.close()\n",
    "    print(\"[*] DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "e7e218f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[*] Producing plot for liquid_mono_culture_orgint_vs_metatranscriptome_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_orgint_vs_metatranscriptome_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_orgint_vs_plate_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_orgint_vs_plate_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_orgint_vs_liquid_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_orgint_vs_hydra_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_orgint_vs_hydra_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for metatranscriptome_vs_liquid_mono_culture_orgint_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for metatranscriptome_vs_liquid_mono_culture_orgint_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for metatranscriptome_vs_plate_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for metatranscriptome_vs_plate_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for metatranscriptome_vs_liquid_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for metatranscriptome_vs_hydra_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_liquid_mono_culture_orgint_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_liquid_mono_culture_orgint_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_metatranscriptome_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_metatranscriptome_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_liquid_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_liquid_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_hydra_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for plate_mono_culture_kiel_vs_hydra_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_kiel_vs_liquid_mono_culture_orgint_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_kiel_vs_metatranscriptome_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_kiel_vs_plate_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_kiel_vs_plate_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_kiel_vs_hydra_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for liquid_mono_culture_kiel_vs_hydra_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_liquid_mono_culture_orgint_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_liquid_mono_culture_orgint_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_metatranscriptome_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_plate_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_plate_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_liquid_mono_culture_kiel_downregulated_genes\n",
      "[*] DONE\n",
      "[*] Producing plot for hydra_mono_culture_kiel_vs_liquid_mono_culture_kiel_upregulated_genes\n",
      "[*] DONE\n"
     ]
    }
   ],
   "source": [
    "for goafile in goafiles:\n",
    "    if isfile(goafile):\n",
    "        goafigure = goafile.split(\"/\")[-1].split(\"_goatools_output.table\")[0]\n",
    "        dataframe = pd.read_table(goafile)\n",
    "        plot_goa(dataframe, result_path, goafigure)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
